<a href="https://colab.research.google.com/github/tec03/continuous_learning/blob/main/non_linear.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# 🛠️ Root-Finding Scenarios

Here we explores scenarios, modeled as nonlinear equations.  

---

## 1. 🔧 Material Behavior (Strain–Stress Relationship)


At what level of material strain does the stress in a forged part reach the critical design threshold of 2 MPa?

Model:    $f(x) = \sqrt{x} - 2$

The root gives the strain level where the material reaches the target stress.

---

## 2. 🔥 Heat Transfer (Thermal Stability)


At what forging temperature do heat generated and heat lost balance perfectly?

Model:  $f(x) = x^3 - 2x - 5$

The root is the steady-state temperature — ideal for thermal process control.

---

## 3. ⚙️ Tool–Workpiece Friction Contact


How much sliding occurs before the tool's frictional force equals the material's resistance?

Model:  $f(x) = \ln(x + 1) - x + 0.5$

The root is the deformation point where forces balance — key for tool life and lubricant decisions.

---

## 4. 🎶 Vibration & Damping


At what point does damping equal restoring force in a vibrating system?

Model:  $f(x) = e^{-x} - \cos(x)$

This root identifies critical displacement — useful in machine stability and noise control.

---

## 5. ❄️ Phase Transition


At what temperature or time does a new phase (solid to liquid) begin?

Model:  $f(x) = x \cdot \sin(x) - 1$

The root gives the onset of phase change — crucial for metallurgical quality.

---

## 6. 🧱 Stress vs Resistance


At what strain does exponential stress overcome linear resistance, risking failure?

Model:  $f(x) = e^x - 3x$

The root marks the critical yield point — vital for load and material safety.

---

Apply numerical methods:
 1. Bisection
 2. False Position
 3. Newton-Raphson
 4. Secant Method

 to solve these problems.

# 1. Bisection Method

## 📘 Bisection Method: Algorithm Steps

1. **Input:**
   - A continuous function $ f(x) $
   - Interval endpoints $ a $ and $ b $
   - Tolerance $\varepsilon$

2. **Check condition:**
   - If $ f(a) \cdot f(b) \geq 0 $, stop. No root is guaranteed in $[a, b]$.

3. **Repeat until convergence:**
   - Compute midpoint: $ c = \frac{a + b}{2} $
   - Evaluate $ f(c) $
   - If $ |f(c)| < \varepsilon $, return $ c $ as the root
   - If $ f(a) \cdot f(c) < 0 $, set $ b = c $; else, set $ a = c $

4. **Stop when:**
   - $ |f(c)| < \varepsilon $, or
   - Maximum number of iterations is reached

5. **Output:**
   - Approximate root $ c $
   - Number of iterations



#### Examples
1. $f(x)=x^3-x-1$
2. $f(x)=2x^3-2x-5$
3. $f(x)=x^3+2x^2+x-1$
4. $f(x) = x^2 - 5$ entre [0, 4]
5. Usted está trabajando para down the toilet company que hace carrozas para inodoros de ABC. La bola flotante tiene una gravedad específica de 0.6 y tiene un radio de 5.5 cm. Se le pide que encuentre la profundidad a la que se sumerge la bola cuando flota en el agua.  La ecuación que da la profundidad x a la cual la pelota está sumergida bajo el agua viene dada por
$$x^3 - 0.165x^2 +3.993x10^{-4} = 0$$
6. $f(x)=x^3+2x^2+x-1$

In [2]:
import math

def bisection_method(func, a, b, tolerance=1e-6, max_iterations=100):
    """
    Finds a root of the function func within the interval [a, b] using the bisection method.

    Returns:
        - Approximate root
        - Number of iterations
        - Iteration details (a, b, midpoint, f(a), f(b), f(midpoint))
    """
    fa = func(a)
    fb = func(b)

    if fa * fb > 0:
        return None, 0, [], "❌ f(a) and f(b) have the same sign — no guaranteed root in this interval."

    iteration_details = []

    for i in range(max_iterations):
        c = (a + b) / 2
        fc = func(c)
        iteration_details.append((i + 1, a, b, c, fa, fb, fc))

        if abs(b - a) < tolerance:
            return c, i + 1, iteration_details, "✅ Root found successfully."

        if fa * fc < 0:
            b = c
            fb = fc
        else:
            a = c
            fa = fc

    return c, max_iterations, iteration_details, "⚠️ Maximum iterations reached."

# ------------------------------------------------------
# Main execution block
# ------------------------------------------------------

try:
    print("Welcome to the Bisection Method Solver!\n")
    user_expr = input("Enter a function in variable x (use * for multiplication, e.g., 2*x**3 - 2*x - 5):\n")
    tolerance = float(input("Enter the desired tolerance (e.g., 1e-6):\n"))
    a = float(input("Enter the left endpoint of the interval (a):\n"))
    b = float(input("Enter the right endpoint of the interval (b):\n"))

    # Define the function safely using eval
    def f(x):
        return eval(user_expr, {"x": x, "math": math, "__builtins__": {}})

    # Run the bisection method
    root, iterations, details, message = bisection_method(f, a, b, tolerance)

    print("\n" + message)

    if root is not None:
        print(f"\n🔎 Approximate root: {root:.8f}")
        print(f"🔁 Number of iterations: {iterations}")
        print("\n📋 Iteration Details:")
        print("Iter |     a     |     b     |   midpoint  |   f(a)   |   f(b)   |  f(midpoint)")
        print("-----|-----------|-----------|-------------|----------|----------|---------------")
        for i, a_val, b_val, c_val, fa, fb, fc in details:
            print(f"{i:>4} | {a_val:>9.4f} | {b_val:>9.4f} | {c_val:>11.4f} | {fa:>8.4f} | {fb:>8.4f} | {fc:>13.4f}")
    else:
        print("Please choose a different interval where f(a) and f(b) have opposite signs.")

except SyntaxError:
    print("\n❌ Syntax error! Did you forget to use '*' for multiplication? (e.g., write '2*x' not '2x')")
except ValueError as ve:
    print(f"\n❌ Value error: {ve}")
except Exception as e:
    print(f"\n❌ Unexpected error: {e}")


Welcome to the Bisection Method Solver!

Enter a function in variable x (use * for multiplication, e.g., 2*x**3 - 2*x - 5):
math.exp(-x)-math.cos(x)
Enter the desired tolerance (e.g., 1e-6):
1e-3
Enter the left endpoint of the interval (a):
1
Enter the right endpoint of the interval (b):
3

✅ Root found successfully.

🔎 Approximate root: 1.29248047
🔁 Number of iterations: 12

📋 Iteration Details:
Iter |     a     |     b     |   midpoint  |   f(a)   |   f(b)   |  f(midpoint)
-----|-----------|-----------|-------------|----------|----------|---------------
   1 |    1.0000 |    3.0000 |      2.0000 |  -0.1724 |   1.0398 |        0.5515
   2 |    1.0000 |    2.0000 |      1.5000 |  -0.1724 |   0.5515 |        0.1524
   3 |    1.0000 |    1.5000 |      1.2500 |  -0.1724 |   0.1524 |       -0.0288
   4 |    1.2500 |    1.5000 |      1.3750 |  -0.0288 |   0.1524 |        0.0583
   5 |    1.2500 |    1.3750 |      1.3125 |  -0.0288 |   0.0583 |        0.0137
   6 |    1.2500 |    1.3125 |   

# 2.Métodos de falsa posición

## 📘 False Position Method (Regula Falsi) — Steps

1. **Input:**
   - Continuous function $ f(x) $
   - Interval $[a, b]$ such that $ f(a) \cdot f(b) < 0 $
   - Tolerance $ \varepsilon $

2. **Check initial condition:**
   - If $ f(a) \cdot f(b) \geq 0 $, stop — no guaranteed root in $[a, b]$

3. **Repeat until convergence:**
   - Compute the new approximation:
     $$c = b - \frac{f(b)(b - a)}{f(b) - f(a)}$$
   - Evaluate $ f(c) $
   - If $ |f(c)| < \varepsilon $, return $ c $ as the root
   - Update the interval:
     - If $ f(a) \cdot f(c) < 0 $, set $ b = c $
     - Else, set $ a = c $

4. **Stop when:**
   - $ |f(c)| < \varepsilon $
   - Or a maximum number of iterations is reached

5. **Output:**
   - Approximate root $ c $
   - Number of iterations


In [3]:
import math

def false_position_method(func, a, b, tolerance=1e-6, max_iterations=100):
    fa = func(a)
    fb = func(b)

    if fa * fb > 0:
        return None, 0, [], "❌ f(a) and f(b) must have opposite signs."

    iteration_details = []

    for i in range(max_iterations):
        c = b - fb * (b - a) / (fb - fa)
        fc = func(c)
        iteration_details.append((i + 1, a, b, c, fa, fb, fc))

        if abs(fc) < tolerance:
            return c, i + 1, iteration_details, "✅ Root found successfully."

        if fa * fc < 0:
            b = c
            fb = fc
        else:
            a = c
            fa = fc

    return c, max_iterations, iteration_details, "⚠️ Maximum iterations reached."


# --- User Input Block ---
try:
    print("Welcome to the False Position Method Solver!\n")
    user_expr = input("Enter a function in x (use * for multiplication, e.g., 2*x**3 - 2*x - 5):\n")
    tolerance = float(input("Enter the desired tolerance (e.g., 1e-6):\n"))
    a = float(input("Enter the left endpoint a:\n"))
    b = float(input("Enter the right endpoint b:\n"))

    def f(x):
        return eval(user_expr, {"x": x, "math": math, "__builtins__": {}})

    root, iterations, details, message = false_position_method(f, a, b, tolerance)

    print("\n" + message)
    if root is not None:
        print(f"\n🔎 Approximate root: {root:.8f}")
        print(f"🔁 Number of iterations: {iterations}")
        print("\n📋 Iteration Details:")
        print("Iter |     a     |     b     |     c     |   f(a)   |   f(b)   |  f(c)")
        print("-----|-----------|-----------|-----------|----------|----------|--------------")
        for i, a_val, b_val, c_val, fa, fb, fc in details:
            print(f"{i:>4} | {a_val:>9.6f} | {b_val:>9.6f} | {c_val:>9.6f} | {fa:>8.5f} | {fb:>8.5f} | {fc:>12.6f}")
    else:
        print("Try using a different interval.")
except SyntaxError:
    print("\n❌ Syntax error: Use * for multiplication (e.g., 2*x not 2x).")
except Exception as e:
    print("\n❌ Error:", e)


Welcome to the False Position Method Solver!

Enter a function in x (use * for multiplication, e.g., 2*x**3 - 2*x - 5):
math.sqrt(x)-2
Enter the desired tolerance (e.g., 1e-6):
1e-3
Enter the left endpoint a:
6
Enter the right endpoint b:
2

✅ Root found successfully.

🔎 Approximate root: 4.00081160
🔁 Number of iterations: 4

📋 Iteration Details:
Iter |     a     |     b     |     c     |   f(a)   |   f(b)   |  f(c)
-----|-----------|-----------|-----------|----------|----------|--------------
   1 |  6.000000 |  2.000000 |  4.263305 |  0.44949 | -0.58579 |     0.064777
   2 |  4.263305 |  2.000000 |  4.037946 |  0.06478 | -0.58579 |     0.009464
   3 |  4.037946 |  2.000000 |  4.005544 |  0.00946 | -0.58579 |     0.001385
   4 |  4.005544 |  2.000000 |  4.000812 |  0.00139 | -0.58579 |     0.000203


# 3. Método de Newton-Raphson

## 📘 Newton-Raphson Method — Steps

1. **Input:**
   - Function $ f(x) $
   - Derivative $ f'(x) $
   - Initial guess $ x_0 $
   - Tolerance $ \varepsilon $

2. **Iterate until convergence:**
   - For each iteration $ n $, compute: $$x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)} $$
   - If $ |f(x_{n+1})| < \varepsilon $, stop and return the root

3. **Stop when:**
   - $ |f(x_{n+1})| < \varepsilon $, or
   - Maximum number of iterations is reached

4. **Output:**
   - Approximate root $ x_{n+1} $
   - Number of iterations
   - Optional: iteration log with $ x_n $, $ f(x_n) $, and $ f'(x_n) $

⚠️ Note:
- Must avoid division by zero $ f'(x_n) = 0 $
- Convergence is **fast** if the guess is good and the function is smooth


In [4]:
import math
from fractions import Fraction

def newton_raphson_method(func, dfunc, x0, tolerance=1e-6, max_iterations=100):
    iteration_details = []

    for i in range(max_iterations):
        fx = func(x0)
        dfx = dfunc(x0)

        if dfx == 0:
            return None, i, iteration_details, "❌ Derivative is zero. Cannot proceed."

        x1 = x0 - fx / dfx
        iteration_details.append((i + 1, x0, fx, dfx, x1))

        if abs(func(x1)) < tolerance:
            return x1, i + 1, iteration_details, "✅ Root found successfully."

        x0 = x1

    return x1, max_iterations, iteration_details, "⚠️ Maximum iterations reached."


# --- User Input Block ---
try:
    print("Welcome to the Newton-Raphson Method Solver!\n")
    user_expr = input("Enter the function f(x) (e.g., x**3 - 2*x - 5):\n")
    deriv_expr = input("Enter its derivative f'(x) (e.g., 3*x**2 - 2):\n")
    x0 = float(Fraction(input("Enter an initial guess: (e.g., 2; or -b/c):\n")))
    tolerance = float(input("Enter the desired tolerance (e.g., 1e-6):\n"))

    # Safe evaluation of functions
    def f(x):
        return eval(user_expr, {"x": x, "math": math, "__builtins__": {}})

    def df(x):
        return eval(deriv_expr, {"x": x, "math": math, "__builtins__": {}})

    root, iterations, details, message = newton_raphson_method(f, df, x0, tolerance)

    print("\n" + message)
    if root is not None:
        print(f"\n🔎 Approximate root: {root:.8f}")
        print(f"🔁 Number of iterations: {iterations}")
        print("\n📋 Iteration Details:")
        print("Iter |    x_n    |   f(x_n)   |  f'(x_n)  |  x_(n+1)")
        print("-----|-----------|------------|-----------|------------")
        for i, x, fx, dfx, x1 in details:
            print(f"{i:>4} | {x:>9.6f} | {fx:>10.6f} | {dfx:>9.6f} | {x1:>10.6f}")
except SyntaxError:
    print("\n❌ Syntax error! Use * for multiplication (e.g., 2*x not 2x).")
except Exception as e:
    print("\n❌ Error:", e)


Welcome to the Newton-Raphson Method Solver!

Enter the function f(x) (e.g., x**3 - 2*x - 5):
math.log(x+1)-x+0.5
Enter its derivative f'(x) (e.g., 3*x**2 - 2):
1/(x+1)-1
Enter an initial guess: (e.g., 2; or -b/c):
2
Enter the desired tolerance (e.g., 1e-6):
1e-6

✅ Root found successfully.

🔎 Approximate root: 1.35767668
🔁 Number of iterations: 3

📋 Iteration Details:
Iter |    x_n    |   f(x_n)   |  f'(x_n)  |  x_(n+1)
-----|-----------|------------|-----------|------------
   1 |  2.000000 |  -0.401388 | -0.666667 |   1.397918
   2 |  1.397918 |  -0.023317 | -0.582972 |   1.357921
   3 |  1.357921 |  -0.000141 | -0.575898 |   1.357677


# 4. Método de secante

## 📘 Secant Method — Steps

1. **Input:**
   - Function $ f(x) $
   - Two initial guesses: $ x_0 $ and $ x_1 $
   - Tolerance $ \varepsilon $

2. **Iterate until convergence:**
   - For each step $ n $, compute: $$x_{n+1} = x_n - f(x_n) \cdot \frac{x_n - x_{n-1}}{f(x_n) - f(x_{n-1})}$$
   - If $ |f(x_{n+1})| < \varepsilon $, stop and return $ x_{n+1} $

3. **Stop when:**
   - $ |f(x_{n+1})| < \varepsilon $, or
   - Maximum number of iterations is reached

4. **Output:**
   - Approximate root
   - Number of iterations
   - Optional: iteration log with $ x_{n-1}, x_n, f(x_n), x_{n+1} $

⚠️ Note: Avoid dividing by zero if $ f(x_n) = f(x_{n-1}) $


In [5]:
import math

def secant_method(func, x0, x1, tolerance=1e-6, max_iterations=100):
    iteration_details = []

    for i in range(max_iterations):
        fx0 = func(x0)
        fx1 = func(x1)

        if fx1 == fx0:
            return None, i, iteration_details, "❌ Division by zero — f(x0) and f(x1) are too close."

        # Secant formula
        x2 = x1 - fx1 * (x1 - x0) / (fx1 - fx0)
        iteration_details.append((i + 1, x0, x1, fx1, x2))

        if abs(func(x2)) < tolerance:
            return x2, i + 1, iteration_details, "✅ Root found successfully."

        # Update guesses
        x0, x1 = x1, x2

    return x2, max_iterations, iteration_details, "⚠️ Maximum iterations reached."


# --- User Input Block ---
try:
    print("Welcome to the Secant Method Solver!\n")
    user_expr = input("Enter the function f(x) (e.g., x**3 - 2*x - 5):\n")
    x0 = float(input("Enter the first initial guess x0:\n"))
    x1 = float(input("Enter the second initial guess x1:\n"))
    tolerance = float(input("Enter the desired tolerance (e.g., 1e-6):\n"))

    # Safe function evaluation
    def f(x):
        return eval(user_expr, {"x": x, "math": math, "__builtins__": {}})

    root, iterations, details, message = secant_method(f, x0, x1, tolerance)

    print("\n" + message)
    if root is not None:
        print(f"\n🔎 Approximate root: {root:.8f}")
        print(f"🔁 Number of iterations: {iterations}")
        print("\n📋 Iteration Details:")
        print("Iter |   x_prev   |    x_curr   |   f(x_curr)   |   x_next")
        print("-----|------------|-------------|---------------|------------")
        for i, x_prev, x_curr, fx_curr, x_next in details:
            print(f"{i:>4} | {x_prev:>10.6f} | {x_curr:>11.6f} | {fx_curr:>13.6f} | {x_next:>10.6f}")
except SyntaxError:
    print("\n❌ Syntax error! Use * for multiplication (e.g., 2*x not 2x).")
except Exception as e:
    print("\n❌ Error:", e)


Welcome to the Secant Method Solver!

Enter the function f(x) (e.g., x**3 - 2*x - 5):
x**3-2*x-5
Enter the first initial guess x0:
3
Enter the second initial guess x1:
0
Enter the desired tolerance (e.g., 1e-6):
1e-4

✅ Root found successfully.

🔎 Approximate root: 2.09455130
🔁 Number of iterations: 8

📋 Iteration Details:
Iter |   x_prev   |    x_curr   |   f(x_curr)   |   x_next
-----|------------|-------------|---------------|------------
   1 |   3.000000 |    0.000000 |     -5.000000 |   0.714286
   2 |   0.000000 |    0.714286 |     -6.064140 |  -3.356164
   3 |   0.714286 |   -3.356164 |    -36.090968 |   1.536343
   4 |  -3.356164 |    1.536343 |     -4.446378 |   2.223789
   5 |   1.536343 |    2.223789 |      1.549586 |   2.046127
   6 |   2.223789 |    2.046127 |     -0.525869 |   2.091142
   7 |   2.046127 |    2.091142 |     -0.037983 |   2.094646
   8 |   2.091142 |    2.094646 |      0.001060 |   2.094551


# 5. Método de Secante Modificada

## 📘 Modified Secant Method — Steps

1. **Input:**
   - Function $ f(x) $
   - Initial guess $ x_0 $
   - Small perturbation factor $ \delta $
   - Tolerance $ \varepsilon $

2. **Repeat until convergence:**
   - Compute: $$x_{n+1} = x_n - \cfrac{f(x_n)}{ \Bigg(\cfrac{f(x_n + \delta x_n) - f(x_n)}{\delta x_n}\Bigg) }$$
   $$x_{n+1} = x_n - \cfrac{f(x_n)* \delta x_n}{ f(x_n + \delta x_n) - f(x_n) } $$
   - If $ |f(x_{n+1})| < \varepsilon $, stop

3. **Stop when:**
   - $ |f(x_{n+1})| < \varepsilon $
   - Or maximum number of iterations is reached

4. **Output:**
   - Approximate root $ x_{n+1} $
   - Number of iterations


In [6]:
import math

def modified_secant_method(func, x0, delta=1e-4, tolerance=1e-6, max_iterations=100):
    iteration_details = []

    for i in range(max_iterations):
        fx = func(x0)
        fx_delta = func(x0 + delta * x0)

        # Approximate derivative
        dfx = (fx_delta - fx) / (delta * x0)

        if dfx == 0:
            return None, i, iteration_details, "❌ Division by zero — approximate derivative is zero."

        # Update x
        x1 = x0 - fx / dfx
        iteration_details.append((i + 1, x0, fx, dfx, x1))

        if abs(func(x1)) < tolerance:
            return x1, i + 1, iteration_details, "✅ Root found successfully."

        x0 = x1  # Update guess

    return x1, max_iterations, iteration_details, "⚠️ Maximum iterations reached."



# --- User Input Block ---
try:
    print("Welcome to the Modified Secant Method Solver!\n")
    user_expr = input("Enter the function f(x) (e.g., x**3 - 2*x - 5):\n")
    x0 = float(input("Enter the initial guess x0:\n"))
    delta = float(input("Enter the small perturbation factor delta (e.g., 1e-4):\n"))
    tolerance = float(input("Enter the desired tolerance (e.g., 1e-6):\n"))

    # Safe evaluation
    def f(x):
        return eval(user_expr, {"x": x, "math": math, "__builtins__": {}})

    root, iterations, details, message = modified_secant_method(f, x0, delta, tolerance)

    print("\n" + message)
    if root is not None:
        print(f"\n🔎 Approximate root: {root:.8f}")
        print(f"🔁 Number of iterations: {iterations}")
        print("\n📋 Iteration Details:")
        print("Iter |    x_n     |   f(x_n)   |  approx f'  |   x_(n+1)")
        print("-----|------------|------------|-------------|------------")
        for i, x, fx, dfx, x1 in details:
            print(f"{i:>4} | {x:>10.6f} | {fx:>10.6f} | {dfx:>11.6f} | {x1:>10.6f}")
except SyntaxError:
    print("\n❌ Syntax error! Use * for multiplication (e.g., 2*x not 2x).")
except Exception as e:
    print("\n❌ Error:", e)


Welcome to the Modified Secant Method Solver!

Enter the function f(x) (e.g., x**3 - 2*x - 5):
math.exp(-x)-math.cos(x)
Enter the initial guess x0:
2
Enter the small perturbation factor delta (e.g., 1e-4):
0.01
Enter the desired tolerance (e.g., 1e-6):
1e-4

✅ Root found successfully.

🔎 Approximate root: 1.29268005
🔁 Number of iterations: 2

📋 Iteration Details:
Iter |    x_n     |   f(x_n)   |  approx f'  |   x_(n+1)
-----|------------|------------|-------------|------------
   1 |   2.000000 |   0.551482 |    0.771085 |   1.284797
   2 |   1.284797 |  -0.005410 |    0.686229 |   1.292680


## ✅ Some **important precautions** $f(x) = \sqrt{x} - 2$



#### ⚠️ Domain Consideration

- The function $ f(x) = \sqrt{x} - 2 $ is only **defined for $ x \geq 0 $**.
- So, if you use a **negative value** as a starting point, the method will **fail** due to a math error (square root of negative).


#### ✅ Safe Use of the Secant Method

Choose two initial values $ x_0 $ and $ x_1 $ such that:

- $ x_0 > 0 $ and $ x_1 > 0 $
- Ideally, $ f(x_0) \cdot f(x_1) < 0 $ so the root lies between them



#### 📍 Example

The root of:

$f(x) = \sqrt{x} - 2$

is clearly:

$x = 4 \quad \text{(since } \sqrt{4} = 2 \text{)}$

Try starting with:

- $ x_0 = 2 \Rightarrow f(2) = \sqrt{2} - 2 \approx -0.586 $
- $ x_1 = 6 \Rightarrow f(6) = \sqrt{6} - 2 \approx 0.45 $

These are valid starting points for the secant method.



#### 🔢 Secant Method Formula

$x_{n+1} = x_n - f(x_n) \cdot \frac{x_n - x_{n-1}}{f(x_n) - f(x_{n-1})}$

Apply this iteratively to get closer to the root.



#### ❌ What Not to Do

Avoid using:

- $ x_0 = -1 $ or $ x_1 = -3 $

because this leads to:

$f(x) = \sqrt{-1} - 2 \quad \text{⛔ Not real-valued}$
